### PR TITLE
Add Gigawords to radius packets where applicable.

### PR DESCRIPTION
**Pull request by jkroonza**
This extends the ppp stats structures to 64 bit unsigned integers.  Due
to the linux kernel (as far as I could determine) only keeps 32 bit
counters in the ppp driver, attempts to count wrap-arounds in the stats
updates.  An improved solution would be to keep 64-bit counters kernel
side and only revert to this if the kernel don't have 64-bit counters.

It further ensures that if interim updates are used that we update the
stats periodically (every 25 seconds if the interim update is longer
than 30 seconds, so as to handle up to Gbps speeds running at maximum
capacity, I chose 25 seconds simply to be somewhat unlikely to collide
with frequently in terms of timing with the actual interim updates
themselves.  This could be anything <30, so a prime may be preferable
(eg, every 29s) but I don't think this matters much.  Without this, it's
possible that we will miss certain wrap-arounds.  This will start
breaking down at speeds >1Gbps.

Signed-off-by: Jaco Kroon <jaco@iewc.co.za>

### CODE REVIEW

**Comment by jkroonza** [comment]
Note:  I just found that my claim above is wrong.

There are 64 bit counters kernel side, they're just being exposed via ioctl() as 32-bit integers.  The problem to expose them via ioctl I *think* is that the ppp_stats structure contains a pppstat structure, prior to the vjstat, and if the size of this structure changes that would represent an ABI breakage on the ioctl(), so one would need to create a new ioctl() with a ppp64stat structure possibly.  I'm not sure what the best approach here would be.  Alternatively, extract from /proc/net/dev or /sys/class/net/${ifname}/statistics/{rx,tx}_bytes.  Happy to write the code for the sysfs mechanism and fall back to ioctl() plus current hack, as this would eliminate the need for polling.  Downside is 6 system calls (open+read+close x 2 for rx+tx) each time the stats is updated, this may be preferable to more frequent ioctl() calls, which my understanding is generally tends to be inefficient in any case due to the table lookup scans kernel side, but I have no reference for a claim that open+read+close on sysfs may be more efficient.  Since interim update must according to RFC be >=60 seconds, we're doing at least 3 ioctl() calls per cycle anyway.

Conceptually thus what we need:  Check for sysfs, if it's available, use it, if not, decide whether or not to periodically poll ioctl() to detect wraps, or revert to old behaviour (doesn't really bug me either way since all my systems has sysfs, and I believe most new deployments should).

**Comment by jkroonza** [comment]
@paulusmack 

Hi Paul, I just finished testing on both branches here, works well in my limited (volume of testing) testing, although, I did try to test properly.  Pushed a link to ~40GB of data and numerous radius updates being sent, and both the _ioctl and _sysfs variants works well.  Would appreciate some form of feedback here.

In our high-volume environments we generally do interim updates every 15 minutes, this means 12 system calls every 15 minutes compared to the previous 1, however to not miss wraps at Gbps on ioctl we need to handle an system call at least every 30 seconds, or about 30 of them over the 15 minute period.  At ~5 minutes things break even, but I did not take the additional overhead for setting up the periodic poller mechanism for ioct() into account, I'm sure this adds a few extra system calls too, so all in all I'm fairly confident the sysfs mechanism shouldn't be worse than the ioctl one overall.

Would appreciate feedback here, happy to make further updates as per direction from yourself.

**Comment by paulusmack** [comment]
Using sysfs to get the 64-bit values is fine with me. Doing a handful of system calls every 25 seconds (or more) should not be a measurable burden on the system.

We could add a new ioctl to the kernel driver, but we would still need the code in pppd to cope with old kernels that don't support the new ioctl, so I think your patch will be needed in any case.

**Comment by pali** [comment]
> Alternatively, extract from /proc/net/dev or /sys/class/net/${ifname}/statistics/{rx,tx}_bytes. Happy to write the code for the sysfs mechanism and fall back to ioctl() plus current hack, as this would eliminate the need for polling. Downside is 6 system calls (open+read+close x 2 for rx+tx) each time the stats is updated, this may be preferable to more frequent ioctl() calls, which my understanding is generally tends to be inefficient in any case due to the table lookup scans kernel side, but I have no reference for a claim that open+read+close on sysfs may be more efficient.

I think you should be able to get these network information (which are available in sysfs or procfs) also via modern rtnetlink socket interface. So it should be only two syscalls (one send() and one blocking recv()) if you prepare and reuse rtnetlink socket.

**Comment by pali** [comment]
Now I'm looking at ppp_generic.c driver and rtnetlink API really provides 64-bit values via ppp_get_stats64() function. So you should really use rtnetlink instead of old ioctl API.

**Comment by jkroonza** [comment]
> Now I'm looking at ppp_generic.c driver and rtnetlink API really provides 64-bit values via ppp_get_stats64() function. So you should really use rtnetlink instead of old ioctl API.

This makes sense, I'll have to figure out how to do rtnetlink though.

**Comment by pali** [comment]
rtnetlink API is mostly undocumented, so you need to look into kernel sources for all details. You can look also at my recent changes in pppd which adds support for setting IPv6 peer address, as it is done via rtnetlink API.

**Comment by jkroonza** [comment]
> rtnetlink API is mostly undocumented, so you need to look into kernel sources for all details. You can look also at my recent changes in pppd which adds support for setting IPv6 peer address, as it is done via rtnetlink API.

Thanks @pali I'll look into this soonest possible.  rtnetlink has always been a total nightmare for me.

**Comment by Neustradamus** [comment]
@jkroonza: Have you looked?

**Comment by jkroonza** [comment]
> @jkroonza: Have you looked?

I have not sorry, been somewhat busy (read: snowed under) on our end with upgrades and new installations.

**Comment by Neustradamus** [comment]
@jkroonza: Have you looked since this summer?

**Comment by jkroonza** [comment]
> @jkroonza: Have you looked since this summer?

Finally done.  I opted to keep the sysfs code as a fall-back to rtnetlink for two reasons:

This is some of the first rtnl code I wrote (have a few bits and pieces of copy&pasted stuff in other projects); and
RTM_GETSTATS was added in 2016 only and I'm not confident it'll be available everywhere.

Also, in comparison with the system calls for radius (from strace), even the sysfs code seems lightweight.

**Review by pali — COMMENTED** [review]
Looks good! See some comments below.

File: pppd/sys-linux.c [hunk 1]
```diff
@@ -140,6 +140,9 @@
 #define NETLINK_CAP_ACK 10
 #endif
 #endif
+#ifdef RTM_GETSTATS
```
[the remark this answers is not in the document]
Comment (pali) [reply pppd/sys-linux.c:143]: Now I checked that support for `RTM_GETSTATS` was added in v4.7-rc1 (commit 10c9ead9f3c6). And support for older kernels is required. There are still officially supported v4.4 and plus enterprise distros have even older supported kernel versions.

[the remark this answers is not in the document]
Comment (pali) [reply pppd/sys-linux.c:143]: And support for `IFLA_STATS64` and `struct rtnl_link_stats64` was added in v2.6.35-rc1 (commit 10708f37ae72).

[the remark this answers is not in the document]
Comment (jkroonza) [reply pppd/sys-linux.c:143]: Wish I had older kernels lying around ... how about this - remove the guarding entirely and if/when someone reports a compile error we figure out those details?

File: pppd/sys-linux.c [hunk 2]
```diff
@@ -1491,6 +1494,118 @@ get_ppp_stats_ioctl(int u, struct pppd_stats *stats)
...
+    if (rtnl_fd < 0) {
+	rtnl_fd = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
+	if (rtnl_fd < 0) {
+	    error("get_ppp_stats_rtnetlink: error creating NETLINK socket: %m (line %d)", __LINE__);
+	    return 0;
+	}
+
+	memset(&nladdr, 0, sizeof(nladdr));
+	nladdr.nl_family = AF_NETLINK;
+	nladdr.nl_pid = getpid();
```
[the remark this answers is not in the document]
Comment (jkroonza) [reply pppd/sys-linux.c:1535]: I noticed in your code.  My testing last night disagreed, but I must've screwed something else up as it works as you state now on my test code block.  Adjusted locally so long.

BTW, reading other examples and stuff, pppd here seems to be makinga few (IMHO reasonable) assumptions about the layout of NL messages.  However, should a compiler decide to pack the structures this will break (eg, from top of memory, struct rtmsg is 12 bytes, but the compiler adds 4 bytes padding to get to the rtnl_link_stats64 structure, on amd64 at least).

There are MACROs in linux/netlink.h and linux/rtnetlink.h to "get to these things the 'correct' way" but my experience with them has also been that it gets messy (lots of pointer handling, and then a macro every time for retrieving the "next" structure - which makes sense if you're to deal with variable sized headers such as IP headers, or even ethernet headers etc where the "next" structure isn't guaranteed to be what you expect...

So I'm in agreement with the current strategy, but it may need some adjustment at some point in the future - so just be aware of it.  I cannot envision that this will change.

File: pppd/sys-linux.c [hunk 3]
```diff
@@ -1491,6 +1494,118 @@ get_ppp_stats_ioctl(int u, struct pppd_stats *stats)
...
+	error("get_ppp_stats_rtnetlink: Netlink responded with error: %s (line %d)", strerror(-nlresp.nlerr.error), __LINE__);
+	goto err;
+    }
+
+    if (nlresp.nlh.nlmsg_type != RTM_NEWSTATS) {
+	error("get_ppp_stats_rtnetlink: Expected RTM_NEWSTATS response, found something else (mlmsg_type %d, line %d)",
+		nlresp.nlh.nlmsg_type, __LINE__);
+	goto err;
+    }
+
```
[the remark this answers is not in the document]
Comment (jkroonza) [reply pppd/sys-linux.c:1595]: NLMSG_OK ... right, looking at the headers what you say makes sense.

So before accessing the sub fields need to check that adequate data has come with the packet (which I'm fairly certain we can assume).  Since the fields we care about are the first four in the stats structure, and the structure in theory can grow over time ... do I just check that the message covers those first four fields, or do I check that we receive the full struct?

```sizeof(struct nlmsg)=16
sizeof(struct rtmsg)=12
sizeof(struct rtnl_link_stats64)=192
sizeof(nlresp)=224
RTM_PAYLOAD(nlresp.nlh) size=196
msg len=224```

Looking at the header for rtnetlink the RTA_OK seems to be the closest thing to NLMSG_OK, and this is for additional attributes not sure it'll fly for this message (There is no rtattr struct).

So I believe the following checks should be adequate:

1.  Before accessing nlmsgerr check that NLMSG_PAYLOAD(&nlh, 0) >= sizeof(nlerr).
2.  Before accessing stats check that RTM_PAYLOAD(&nlh) >= sizeof(stats).

Two works because RTM_PAYLOAD is defined in terms of NLMSG_PAYLOAD.

File: pppd/sys-linux.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkroonza) [reply pppd/sys-linux.c:143]: Ended up just protecing ```RTM_GETSTATS``` and ```IFLA_STATS_LINK_64```.

I have bumped into a 2.6.32 kernel not too long ago, but I don't see an easy way of protecting IFLA_STATS64 and specifically the rtnl_link.  Best approach there is to probably not compile in the support if kernel headers indicates it's older than 2.6.35.

File: pppd/sys-linux.c [hunk 4]
```diff
@@ -1491,6 +1494,118 @@ get_ppp_stats_ioctl(int u, struct pppd_stats *stats)
...
+    if (!NLMSG_OK(&nlresp.nlh, nlresplen)) {
+	error("get_ppp_stats_rtnetlink: Netlink response message is corrupt/incomplete (line %d)", __LINE__);
+	goto err;
+    }
+
+    if (nlresp.nlh.nlmsg_type == NLMSG_ERROR) {
+	error("get_ppp_stats_rtnetlink: Netlink responded with error: %s (line %d)", strerror(-nlresp.nlerr.error), __LINE__);
```
[the remark this answers is not in the document]
Comment (jkroonza) [reply pppd/sys-linux.c:1586]: It'll only crop up once, and would be the only indication that we're falling back to sysfs stats, so ideally I'd think keep it.

File: pppd/sys-linux.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (pali) [reply pppd/sys-linux.c:1595]: Check for first four fields should be enough. Also it should be possible to instruct kernel to just send to userspace only truncated buffer with these first four fields.

File: pppd/sys-linux.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (pali) [reply pppd/sys-linux.c:143]: I edited https://github.com/ppp-project/ppp/pull/298#discussion_r765300431 and put there a new version of possible guard. It is untested, but probably something like that could work.

[the remark this answers is not in the document]
Comment (jkroonza) [reply pppd/sys-linux.c]: Ok, so here is the thing:

```607ca46e97a1b (David Howells                  2012-10-13 10:46:48 +0100  215) struct rtnl_link_stats64 {```

No changes to this structure was made (except to formalize the comments) since.

I don't believe we can request only a portion of the stats, and (NLMSG_OK() would start failing if we only receive a partial packet ... might be worthwhile to actually pad the structure on our end a bit in any case to ensure that we get the whole thing).  Or I need to re-evaluate the use of that MACRO to permit only a portion, but eventually I'm going to have to RX the whole thing anyway since I believe follow-up recv's would fetch the tail of the last kernel-side tx.  Which brings me back to possible need to allocate a slightly larger buffer on the stack so that we don't run into this, although, in such cases I do simply close the socket and fail (which will fall back to sysfs, but if we increase the buffer size we could proceed even if the size of the structure does change for some unforeseen reason).

@pali if you could point me at something that says something different that would be great.

File: pppd/sys-linux.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (pali) [reply pppd/sys-linux.c]: My concern is that this is _error_ message. I do not think that we should spam error part of syslog (or any other system log) about this error on system which do not support RTM_GETSTATS. On these systems it could be just informational note message.

Error messages in system log lot of times trigger alerts.

See how it is done in `sif6addr_rtnetlink` function. On older systems there this particular information (about not supported functionality) is not shown and pppd just fallbacks to older implementation.

File: pppd/sys-linux.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkroonza) [reply pppd/sys-linux.c]: So I did this as a test (in a completely different program):

``` 13 struct rtnl_link_stats64_compat {
 14   uint64_t rx_packets;
 15   uint64_t tx_packets;
 16   uint64_t rx_bytes;
 17   uint64_t tx_bytes;
 18 };
 19 #define rtnl_link_stats64 rtnl_link_stats64_compat ```
 
 ```141     struct {
142         struct nlmsghdr nlh;
143         union {
144             struct nlmsgerr err;
145             struct {
146                 struct rtmsg rth;
147                 struct rtnl_link_stats64 stats;
148             };
149         };
150     } resp;```

```202     while (1) {
203         printf("Sending request\n");
204         if (sendmsg(fd, &req_msg, 0) < 0) {
205             perror("sendmsg");
206             return 1;
207         }
208 
209         printf("Receiving message\n");
210         resp_size = recvmsg(fd, &resp_msg, 0);
211         if (resp_size < 0) {
212             perror("recvmsg");
213             return 1;
214         }
215         printf("MSG size: %ld\n", resp_size);
216 
217         if (!NLMSG_OK(&resp.nlh, resp_size)) {
218             printf("NLMSG is NOT OK\n");
219             return 1;
220         }
221 
222         printf("NL sub message: %d\n", resp.nlh.nlmsg_type);
223         switch (resp.nlh.nlmsg_type) {
224         case NLMSG_ERROR:
225             printf("Error code: %d (%s)\n", resp.err.error, strerror(-resp.err.error));
226             goto sleep;
227         case RTM_NEWSTATS:
228             printf("Got stats.\n");
229 #define outp(x) printf("%20s: %5u\n", #x, (unsigned)resp.rth.x)
230             outp(rtm_family);
231             outp(rtm_dst_len);
232             outp(rtm_src_len);
233             outp(rtm_tos);
234             outp(rtm_table);
235             outp(rtm_protocol);
236             outp(rtm_scope);
237             outp(rtm_type);
238             outp(rtm_flags);
239 #undef outp
240 #define outp(x) printf("%20s: %20lu\n", #x, resp.stats.x)
241             outp(rx_packets);
242             outp(tx_packets);
243             outp(rx_bytes);
244             outp(tx_bytes);
245 //          outp(rx_errors);
246 //          outp(tx_errors);
247 //          outp(rx_dropped);
248 //          outp(rx_dropped);
249 #undef outp
250             break;
251         default:
252             printf("Unknown sub type (%d).\n", resp.nlh.nlmsg_type);
253         }
254 
255 sleep:
256         sleep(10);
257     }```

Tried to trim things down as much as possible, I set up a single req_msg for sending every 10 seconds.  Not much in the way of testing here.  NLMSG_OK fails instantly:

```sizeof resp: 64
Sending request
Receiving message
MSG size: 64
NLMSG is NOT OK```

Eliminating the NLMSG_OK and replacing with:

```227         if (resp_size < (int)sizeof(resp.nlh)) {
228             printf("response is smaller than nlh\n");
229             return 1;
230         }```

then just manually checking as we descend that we retrieved at least the bytes required (without using NLMSG_PAYLOAD since that's just retrieving the fields in the nlmsg structure so can't rely on them due to NLMSG_OK being false - we didn't retrieve the whole message), so something like:

```142     struct resp_structure {
143         struct nlmsghdr nlh;
144         union {
145             struct {
146                 struct nlmsgerr err;
147                 char end_err[0];
148             };
149             struct {
150                 struct rtmsg rth;
151                 struct rtnl_link_stats64 stats;
152                 char end_stats[0];
153             };
154         };
155     } resp;
...
242             if (resp_size < (int)offsetof(struct resp_structure, end_stats)) {
243                 printf("Incomplete stats\n");
244                 return 1;
245             }```

Where offsetof is used before accessing the statistics in this case, similar checks for error messages ... so we just can't rely on the MACROs from netlink.h.  Happy to make this adjustments if you prefer.  I think the code just gets nastier and harder to read in order to save a few bytes being copied between kernel/user space ...

File: pppd/sys-linux.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (jkroonza) [reply pppd/sys-linux.c]: Point taken.  Are we certain about the version being 4.7.0?  Guessing checking against EOPNOTSUPPORT is the more generic test, but added as requested.

File: pppd/sys-linux.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (jkroonza) [reply pppd/sys-linux.c:1602]: Done and as side effect the messages from kernel space is not truncated too ...

File: pppd/sys-linux.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (pali) [reply pppd/sys-linux.c]: Perfect testing! This is probably reason why I also have not used `NLMSG_OK` macro...

File: pppd/sys-linux.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (pali) [reply pppd/sys-linux.c]: Version 4.4 is LTS supported version and is older than 4.7.0.

File: pppd/sys-linux.c [hunk 5]
```diff
@@ -127,9 +128,27 @@
 #include <linux/ppp_defs.h>
 #include <linux/if_ppp.h>
 
-#ifdef INET6
 #include <linux/netlink.h>
 #include <linux/rtnetlink.h>
+#include <linux/if_link.h>
+/* Attempt at retaining compile-support with older than 4.7 kernels,
+ * older than 2.6.35 will probably break.  Either way, pre 4.7 kernels
+ * will not work with this code. */
```
Comment (pali) [inline pppd/sys-linux.c:136]: Just to note that this ifdef is compile time dependency, not runtime dependency. It is possible to use reduced header files or SDK (probably in embedded or cross compile toolchains) where not all constants are defined. So this can be required also when using new post 4.7 kernels. IIRC people already reported issues that pppd compilation failed just because they did not have `NETLINK_CAP_ACK` or `SOL_NETLINK` in header files.

File: pppd/sys-linux.c [hunk 6]
```diff
@@ -127,9 +128,27 @@
...
+#define IFLA_STATS_LINK_64 1
+#endif
+/* Since we only care about the initial few bytes, we use our own structure */
+struct rtnl_link_stats64_compat {
+    uint64_t rx_packets;
+    uint64_t tx_packets;
+    uint64_t rx_bytes;
+    uint64_t tx_bytes;
+};
+#define rtnl_link_stats64 rtnl_link_stats64_compat
```
Comment (pali) [inline pppd/sys-linux.c:149]: If you are not going to hide/guard this custom definition of `rtnl_link_stats64` then rather do not define it at all. And instead put it directly into `struct nlresp`. If it is possible it is better to not redefine or overwrite _standard_ structures to prevent possible future breakage. My original idea was to hide this `struct rtnl_link_stats64_compat` just for toolchains which are older (or do not define all constants) as these toolchains have less probability of breakage.

**Comment by pali** [comment]
Just one github related comment. Avoid putting person name prefixed with `@` character in commit messages. Because github has totally insane behavior as it on every push on such commit sends email to marked person... And it happens for every repository fork, which results in email bomb by github.

And if you want to look at another netlink code, I have prepared (but not finished yet) support for registering ppp interface via Linux rtnetlink API. It is in my repository fork in register-with-name branch. (I totally forgot that I have prepared it months ago).

File: pppd/sys-linux.c (hunk 5 above)
Comment (jkroonza) [reply pppd/sys-linux.c:136]: That comment is broken either way, since now 2.6.35 and older should compile too, so comment updated.

File: pppd/sys-linux.c (hunk 6 above)
Comment (jkroonza) [reply pppd/sys-linux.c:149]: Updated.

**Comment by jkroonza** [comment]
I marked resolved the conversations I believe to be sorted out.  Please do confirm, but I believe this should now be all sorted.  Including the previously missing Signed-off-by header (which I'm not sure ppp requires but I added regardless).

File: pppd/sys-linux.c [hunk 7]
```diff
@@ -1469,9 +1476,131 @@ get_ppp_stats(int u, struct pppd_stats *stats)
...
+{
+    struct pppd_stats dummy;
+    get_ppp_stats_ioctl((long)u, &dummy);
+    TIMEOUT(ppp_stats_poller, u, 25);
+}
+
+/********************************************************************
+ * get_ppp_stats - return statistics for the link.
+ */
+int get_ppp_stats(int u, struct pppd_stats *stats)\
```
Comment (pali) [inline pppd/sys-linux.c:1725]: Trailing backslash.

File: pppd/sys-linux.c [hunk 8]
```diff
@@ -1469,9 +1476,131 @@ get_ppp_stats(int u, struct pppd_stats *stats)
...
+	    error("%s: %m", fname);
+	    return 0;
+	}
+	/* trim trailing \n if present */
+	while (rlen > 0 && buf[rlen-1] == '\n')
+	    rlen--;
+	buf[rlen] = 0;
+
+	val = strtoull(buf, &err, 10);
+	if (err && *err) {
```
Comment (pali) [inline pppd/sys-linux.c:1686]: This is incorrect error checking if `strtoull` failed. `strto*` has lot of traps, look into [documentation](https://man7.org/linux/man-pages/man3/strtol.3.html). You have to check also for `errno` and validity of the first byte in `buf`.

File: pppd/main.c [hunk 9]
```diff
@@ -1240,9 +1241,9 @@ update_link_stats(int u)
 
     slprintf(numbuf, sizeof(numbuf), "%u", link_connect_time);
     script_setenv("CONNECT_TIME", numbuf, 0);
-    slprintf(numbuf, sizeof(numbuf), "%u", link_stats.bytes_out);
+    snprintf(numbuf, sizeof(numbuf), "%" PRIu64, link_stats.bytes_out);
     script_setenv("BYTES_SENT", numbuf, 0);
-    slprintf(numbuf, sizeof(numbuf), "%u", link_stats.bytes_in);
+    snprintf(numbuf, sizeof(numbuf), "%" PRIu64, link_stats.bytes_in);
```
Comment (pali) [inline pppd/main.c:1246]: I'm not sure if replacement slprintf by snprintf is ok. @paulusmack should verify it.

**Review by pali — COMMENTED** [review]
Second netlink commit is fine! Now I have looked at the first commit and below are some comments.

File: pppd/main.c (hunk 9 above)
Comment (jkroonza) [reply pppd/main.c:1246]: slprintf doesn't work in this case.

File: pppd/sys-linux.c (hunk 8 above)
Comment (jkroonza) [reply pppd/sys-linux.c:1686]:
```
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <errno.h>

int main(int, char **argv)
{
	unsigned long long r;
	char *endptr;

	while (*++argv) {
		r = strtol(*argv, &endptr, 0);
		printf("i=%p\"%s\" r=%llu, errno=%d, endptr=%p\"%s\"(%lu)\n",
				*argv, *argv, r, errno, endptr, endptr, strlen(endptr));
	}
	return 0;
}
```

Now this is interesting (ERANGE==34):

```
jkroon@plastiekpoot ~/src/test $ ./strtol 01a "" "-1" "18446744073709551616" 0 1
i=0x7ffdcd69bfaf"01a" r=1, errno=0, endptr=0x7ffdcd69bfb1"a"(1)
i=0x7ffdcd69bfb3"" r=0, errno=0, endptr=0x7ffdcd69bfb3""(0)
i=0x7ffdcd69bfb4"-1" r=18446744073709551615, errno=0, endptr=0x7ffdcd69bfb6""(0)
i=0x7ffdcd69bfb7"18446744073709551616" r=18446744073709551615, errno=34, endptr=0x7ffdcd69bfcb""(0)
i=0x7ffdcd69bfcc"0" r=0, errno=34, endptr=0x7ffdcd69bfcd""(0)
i=0x7ffdcd69bfce"1" r=1, errno=34, endptr=0x7ffdcd69bfcf""(0)
```

I'd expect -1 to also set errno=ERANGE (and return 0, but ULONG_MAX is in line with documentation).  I'd expect strtoll (if it's going to use errno for returning errors) to zero it upon entry.  "lots of traps" indeed.  Hopeing it's covered now.

Comment (pali) [reply pppd/sys-linux.c]: Yes, `strtou...` functions do not signal underflow. This is really a trap. Anyway, you do not need to test `err` pointer for non-NULL. After successful call (when errno == 0) is `err` set, so you can dereference `*err`.

File: pppd/sys-linux.c [hunk 10]
```diff
@@ -1469,7 +1487,262 @@ get_ppp_stats(int u, struct pppd_stats *stats)
...
+	    return 0;
+	}
+	/* trim trailing \n if present */
+	while (rlen > 0 && buf[rlen-1] == '\n')
+	    rlen--;
+	buf[rlen] = 0;
+
+	errno = 0;
+	val = strtoull(buf, &err, 10);
+	if (*buf < '0' || *buf > '9' || errno != 0 || err && *err) {
```
Comment (pali) [inline pppd/sys-linux.c:1687]: Mixing `| |` and `&&` operators in one condition without putting sub-conditions in parenthesis is really a bad idea.

File: pppd/sys-linux.c (hunk 8 above)
Comment (jkroonza) [reply pppd/sys-linux.c]: Dropped the check for err being non-null.

File: pppd/sys-linux.c (hunk 10 above)
Comment (jkroonza) [reply pppd/sys-linux.c]: With check for err being non-null dropped, it's all ||.

That said, mathematically and in every compiler I've ever worked with, && has higher priority than ||, so I have to respectfully disagree in that there is no ambiguity here regardless.  Unless some compiler somewhere handles this incorrectly or there is something else I'm not aware of.

Comment (pali) [reply pppd/sys-linux.c]: Compilers handle it correctly and you are right that this is mathematically correct. But compilers throw warnings about such usage and it should be avoided in new code. E.g. new versions of gcc have warning category `-Wparentheses` which is automatically enabled by `-Wall`. And using `-Wall` is really recommended. If you disagree with it then you would have to consult it with gcc developers.

Comment (jkroonza) [reply pppd/sys-linux.c]: Right.  Avoid compiler warning.  Understood.  Thanks for the explanation, appreciated.

**Comment by Neustradamus** [comment]
@jkroonza: @RICCIARDI-Adrien has commented here: https://github.com/jkroonza/ppp/commit/a21a280ad1cc0d4b6551d5196622435d20258022#diff-42a0dd69046ea1e7a1aa82bbec76ae7c90eaf4dea269e9cb779701cbd045ad4bR178

What do you think?

**Comment by jkroonza** [comment]
Simply rebased to get the solaris build fixes from master to hopefully pass CI now.

**Comment by paulusmack** [comment]
I squashed the two commits into one because there were was more in the second commit than just adding the netlink stuff. For example there was the u_int64_t to uint64_t change, and an unnecessary backslash that was added in the first commit and removed in the second.

**Comment by jkroonza** [comment]
> I squashed the two commits into one because there were was more in the second commit than just adding the netlink stuff. For example there was the u_int64_t to uint64_t change, and an unnecessary backslash that was added in the first commit and removed in the second.

Thank you!  Would have been happy to handle that on my end, and yea, I should have put that in the first commit ... appreciate the assistance and guidance from all involved.
